aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +685,513 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+
+  auto Bs = at::native::_compute_linear_combination(As, bs);
+
+  // compute A9
+  Bs.select(0, 3).add_(at::matmul(Bs.select(0, 0), Bs.select(0, 4)));
+  return Bs.select(0, 1).add_(at::matmul(
+    Bs.select(0, 2).add_(Bs.select(0, 3)),
+    Bs.select(0, 3)
+  ));
+}
+
+template <typename scalar_t>
+void compute_T18_scale_square(
+  Tensor& mexp_out,
+  const Tensor& a,
+  const Tensor& norm,
+  scalar_t theta
+) {
+  // Scale
+  const auto s = at::max(
+    at::zeros_like(norm),
+    at::ceil(at::log2(norm / theta))
+  ).unsqueeze(-1).unsqueeze(-1).to(at::kLong);
+  const auto pow2s = at::pow(2, s);
+  const auto a_scaled = a / pow2s;
+
+  // Square
+  auto mexp_scaled = at::native::compute_T18<scalar_t>(a_scaled);
+  auto s_cpu = s.to(at::kCPU);
+  for (int64_t i = 0; i < mexp_scaled.size(0); ++i) {
+    auto s_val = s_cpu.select(0, i).template item<int64_t>();
+    auto mexp = mexp_scaled.select(0, i);
+    for (int64_t p = 0; p < s_val; ++p) {
+      mexp = at::matmul(mexp, mexp);
+    }
+    mexp_out.select(0, i).copy_(mexp);
+  }
+}
+
+template <typename scalar_t>
+Tensor mexp_impl(
+  const Tensor& a,
+  std::array<scalar_t, total_n_degs> thetas,
+  bool compute_highest_degree_approx = false
+) {
+  auto res = at::empty_like(a);
+  const auto norm = operator_1_norm(a);
+  const auto norm_cpu = norm.to(at::kCPU);
+
+  if (!compute_highest_degree_approx) {
+    constexpr std::array<
+      Tensor(*)(const Tensor&),
+      total_n_degs - 1> 
+    compute_Ts = {
+      compute_T1, compute_T2, compute_T4<scalar_t>,
+      compute_T8<scalar_t>, compute_T12<scalar_t>
+    };
+
+    for (int i = 0; i < total_n_degs - 1; ++i) {
+      auto norm_lower_bound = (i == 0) ? static_cast<scalar_t>(-1) : thetas[i - 1];
+      auto norm_upper_bound = thetas[i];
+      // nonzero returns a 2D tensor, hence squeeze(-1) to make it 1D
+      auto idx_curr_norm_interval = (
+        (norm_lower_bound < norm_cpu) * (norm_cpu <= norm_upper_bound)
+      ).nonzero().squeeze(-1);
+
+      if (idx_curr_norm_interval.numel()) {
+        // We pin memory to make the transfer to CUDA faster and async
+        auto idx_to_device = _pin_and_move_memory_if_cuda_input(
+          idx_curr_norm_interval, a
+        );
+        auto sub_a = at::index_select(a, 0, idx_to_device);
+        res.index_put_({idx_to_device}, compute_Ts[i](sub_a));
+      }
+    }
+
+    // nonzero returns a 2D tensor, hence squeeze(-1) to make it 1D
+    auto idx_large_norm = (norm_cpu >= thetas[total_n_degs - 2])
+      .nonzero().squeeze(-1);
+
+    if (idx_large_norm.numel()) {
+      // We pin memory to make the transfer to CUDA faster and async
+      auto idx_to_device = _pin_and_move_memory_if_cuda_input(
+        idx_large_norm, a
+      );
+      auto a_large_norm = at::index_select(a, 0, idx_to_device);
+      auto large_norm_subset = at::index_select(norm, 0, idx_to_device);
+      auto mexp_out = at::empty_like(a_large_norm);
+
+      compute_T18_scale_square(
```

Comment:
It seems a bit unfortunate that this part of the calculation (T18 expansion) can't be done uniformly as part of the same loop above.

Reply:
E.g., maybe you could just stick one last entry in `thetas` that is Inf?

Reply:
Unfortunately not, the approximation schemes are only effective for inputs close to zero, so the higher degree requires scaling for inputs with large norms ;(

Reply:
> E.g., maybe you could just stick one last entry in `thetas` that is Inf?

Need to perform scaling in order for T18 be a proper approximation.